src/Interfaces/AbstractHPolygon.jl

```diff
@@ -543,39 +541,49 @@ Perform a binary search in the constraints.
 
 - `d`            -- direction
 - `constraints`  -- constraints
-- `n`            -- number of constraints
-- `k`            -- start index
+- `start_index`  -- (optional, default: `div(length(constraints)+1, 2)`) start
+                    index
 - `choose_lower` -- (optional, default: `false`) flag for choosing the lower
                     index (see the 'Output' section)
 
 ### Output
 
 In the default setting, the result is the smallest index `k` such that
-`d <= constraints[k]`, or `n+1` if no such `k` exists.
+`d <= constraints[k].a`, or `length(constraints)+1` if no such `k` exists.
 If the `choose_lower` flag is set, the result is the largest index `k` such
-that `constraints[k] < d`, which is equivalent to being `k-1` in the normal
+that `constraints[k].a < d`, which is equivalent to being `k-1` in the normal
 setting.
 """
 function binary_search_constraints(d::AbstractVector{N},
-                                   constraints::Vector{<:HalfSpace{N}},
-                                   n::Int,
-                                   k::Int;
+                                   constraints::Vector{<:HalfSpace{N}};
+                                   start_index::Int=div(length(constraints)+1, 2),
                                    choose_lower::Bool=false) where {N}
     lower = 1
-    upper = n+1
+    n = length(constraints)
+    upper = n + 1
+    @assert 1 <= start_index <= n "invalid start index $start_index"
+    m = start_index
     while lower + 1 < upper
-        if constraints[k].a <= d
-            lower = k
+        if constraints[m].a <= d
+            lower = m
         else
-            upper = k
+            upper = m
         end
-        k = lower + div(upper - lower, 2)
+        m = div(lower + upper, 2)
     end
+
+    # since `<=` is approximate, it can happen that x <= y and y <= x
+    # we want to return the smallest index, so (linearly) search to the left
+    while lower > 1 && d <= constraints[lower].a && constraints[lower].a <= d
```

Comment:
can't we use `_isaprox` here?

Reply:
Here is a counterexample:

```julia
julia> u = ones(2);
julia> v = u .+ 1e-5;

julia> _isapprox(u, v)
false

julia> u <= v <= u
true
```

---

`u <= v` (defined for 2D vectors) falls back to `_leq_quadrant(u, v)`, which checks if the vectors are in different quadrants. If they are not, it returns `is_right_turn(u, v)`, which is `_geq(right_turn(u, v), 0)`. And `right_turn(u, v)` is `u[1] * v[2] - u[2] * v[1]`.

So `u <= v <= u` holds if `_geq(u[1] * v[2] - u[2] * v[1], 0) && _geq(v[1] * u[2] - v[2] * u[1], 0)`.

Not sure this helps (it did not for me) :sweat_smile: